
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="datatree.html">
      
      
        <link rel="next" href="sequential.html">
      
      
      <link rel="icon" href="images/gsas2.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.23">
    
    
      
        <title>Universal items - GSAS-II Help</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.84d31ad4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#universal-data-tree-items" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="GSAS-II Help" class="md-header__button md-logo" aria-label="GSAS-II Help" data-md-component="logo">
      
  <img src="images/gsas2.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GSAS-II Help
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Universal items
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="GSAS-II Help" class="md-nav__button md-logo" aria-label="GSAS-II Help" data-md-component="logo">
      
  <img src="images/gsas2.png" alt="logo">

    </a>
    GSAS-II Help
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contents
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="preface.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="applicationwindow.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    GUI components
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="mainmenu.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Main menu
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Data tree items
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Data tree items
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="datatree.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data tree overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Universal items
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="commontreeitems.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Universal items
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#notebook" class="md-nav__link">
    <span class="md-ellipsis">
      Notebook
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#controls" class="md-nav__link">
    <span class="md-ellipsis">
      Controls
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#covariance" class="md-nav__link">
    <span class="md-ellipsis">
      Covariance
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#constraints" class="md-nav__link">
    <span class="md-ellipsis">
      Constraints
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Constraints">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequential-refinement-constraints" class="md-nav__link">
    <span class="md-ellipsis">
      Sequential Refinement Constraints
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#restraints" class="md-nav__link">
    <span class="md-ellipsis">
      Restraints
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Restraints">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#restraint-types" class="md-nav__link">
    <span class="md-ellipsis">
      Restraint types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#importing-restraints" class="md-nav__link">
    <span class="md-ellipsis">
      Importing Restraints
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#restraint-weights" class="md-nav__link">
    <span class="md-ellipsis">
      Restraint Weights
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rigid-bodies" class="md-nav__link">
    <span class="md-ellipsis">
      Rigid bodies
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="sequential.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sequential results
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="cluster.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cluster Analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5" >
        
          
          <label class="md-nav__link" for="__nav_5_5" id="__nav_5_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Phases
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5">
            <span class="md-nav__icon md-icon"></span>
            Phases
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phaseoverview.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Phase overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasegeneral.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    General tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasedata.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phaseatoms.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Atoms tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasedrawopts.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Draw Options tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasedrawatoms.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Draw Atoms tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phaseRB.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RB Models tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasetexture.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Texture tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasemappeaks.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Map Peaks tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasepawley.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Pawley tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phaselayers.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Layers tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasewave.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Wave Data tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasemcsa.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    MC/SA tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phaseRMC.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    RMC tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phaseisodistort.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    ISODISTORT tab
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="phasedysnomia.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Dysnomia tab
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="image.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Images (IMG)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_7" >
        
          
          <label class="md-nav__link" for="__nav_5_7" id="__nav_5_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Powder (PWDR)
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_7">
            <span class="md-nav__icon md-icon"></span>
            Powder (PWDR)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powder.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderparent.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Main PWDR entry
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powdercomments.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Comments
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderlimits.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Limits
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderbkg.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Background
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderinst.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Instrument Parameters
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powdersample.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sample Parameters
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderpeaks.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Peak List
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderindexppeaks.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Index Peak List
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powdercells.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Unit Cells List
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="powderrefs.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reflection Lists
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="singlecrystal.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Single Crystal (HKLF)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="pairdistribution.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    PDF
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="smallanglescattering.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Small Angle (SASD)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reflectometry.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reflectometry (REFD)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="peaks.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Powder Peaks (PKS)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="others.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Other info
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#notebook" class="md-nav__link">
    <span class="md-ellipsis">
      Notebook
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#controls" class="md-nav__link">
    <span class="md-ellipsis">
      Controls
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#covariance" class="md-nav__link">
    <span class="md-ellipsis">
      Covariance
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#constraints" class="md-nav__link">
    <span class="md-ellipsis">
      Constraints
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Constraints">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequential-refinement-constraints" class="md-nav__link">
    <span class="md-ellipsis">
      Sequential Refinement Constraints
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#restraints" class="md-nav__link">
    <span class="md-ellipsis">
      Restraints
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Restraints">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#restraint-types" class="md-nav__link">
    <span class="md-ellipsis">
      Restraint types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#importing-restraints" class="md-nav__link">
    <span class="md-ellipsis">
      Importing Restraints
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#restraint-weights" class="md-nav__link">
    <span class="md-ellipsis">
      Restraint Weights
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rigid-bodies" class="md-nav__link">
    <span class="md-ellipsis">
      Rigid bodies
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<!--- Don't change the HTML version of this file; edit the .md version -->
<h1 id="universal-data-tree-items">Universal Data Tree Items</h1>
<p>The data tree items listed below will appear in every GSAS-II project once any histogram or phase has been placed in the project. </p>
<p><a name="Notebook"></a></p>
<h2 id="notebook">Notebook</h2>
<p>This window provides a log of information on what has been done in a project.
Some GSAS-II operations (<em>e.g.</em>, structure refinement, Fourier map calculation &amp; peak fitting) will add entries to the notebook. You may also place whatever text commentary you wish into the window by clicking on the "Add comment" button. The comment will be time-stamped and placed appropriate order in the log. </p>
<ul>
<li>
<p>Normally entries are shown in reverse chronological order ("newest-1st"), but that button allows sorting with the oldest entries first ("oldest-1st").</p>
</li>
<li>
<p>The "Set filters" button allows you to select which type of entries are shown, as are selected on a list. Note that older Notebook entries were not recorded with tags that allow filtering to work. </p>
</li>
<li>
<p>The "Plot" button allows plotting the progression of the overall Rw or reduced <span class="arithmatex">\(\chi^2\)</span> in the refinement. </p>
</li>
<li>
<p>The "Save" button writes the contents of the Notebook as a text file named &lt;project&gt;-notebook.txt. </p>
</li>
</ul>
<H3 style="color:blue;font-size:1.1em">What can I do here?</H3>

<p>Use the notebook to keep track of information related to how you use GSAS-II, what has been done or to plot the progress of the fit. When you make different copies of a fit to try out different refinement options, it will be very helpful to put a comment here to help you remember what you did. </p>
<p><a name="Controls"></a></p>
<h2 id="controls">Controls</h2>
<p>This window provides access to the controls that determine how GSAS-II performs minimizations as well as few global parameters for GSAS-II. Note that many other customization settings are set as  <a href="others.html#config">configuration variables</a> in the File/Preferences menu. (See the 
<a href="https://gsas-ii.readthedocs.io/en/latest/GSASIIutil.html#config-example-py-configuration-options">Configuration variables section of the Programmer's documentation.</a> for listing of these options.)</p>
<ol>
<li>
<p><strong>Refinement Controls</strong>: 
<a name="RefinementControls"></a>
 These controls determine how refinements are performed. The first determines the computational engine used to minimize the structure.</p>
<ul>
<li><strong>Refinement type</strong> - 
<a name="RefineType"></a>
This determines how structure refinements are performed. The choices are:<ul>
<li><strong>analytic Hessian</strong>: This is the default option and is usually the most useful. It uses a custom-developed least-squares minimizer that uses singular-value decomposition (SVD) to reduce the errors caused by correlated variables and the Levenberg-Marquardt algorithm to up-weight diagonal Hessian terms when a refinement step fails to lower χ2.</li>
<li><strong>analytic Jacobian</strong>: This uses the numpy-provided <code>leastsq</code> minimizer, which not applicable for problem with a large number of histograms as it requires much more memory than the Hessian routines. This because it creates a Jacobian matrix (J) that is shaped N x M (N parameters x M observations) that is used to create the N x N Hessian. The Hessian method creates a Jacobian matrix only for each histogram; the N x N Hessian is the made from summing the J x J<span class="arithmatex">\(^T\)</span> products across the histograms</li>
<li><strong>numeric</strong>: This also uses the numpy leastsq minimizer and is also not applicable for larger problems. Unlike, the "analytic Jacobian", numerical derivatives are computed rather than use the analytical derivatives that are coded directly into GSAS-II. This will be slower than the analytical derivatives and will is often less accurate which results in slower convergence. It is typically used for code development to check the accuracy of the analytical derivative formulations.</li>
<li><strong>Hessian SVD</strong>: This is very similar to analytic Hessian but does not include the Levenberg-Marquardt algorithm. It can be faster but is more prone to diverge when severe correlation is present. It is possible that it might be better for single-crystal refinements.</li>
</ul>
</li>
</ul>
<p>Note that the Jacobian refinement tools are the Fortran MINPACK<code>lmdif</code> and <code>lmder</code> algorithms wrapped in Python, as provided in the numpy/Scipy package. The Hessian routines were developed for GSAS-II based on routines in numpy and scipy and used aw well material from <em>Numerical Recipes</em> (Press, Flannery, Teulosky &amp; Vetterling) for the Levenberg-Marquardt algorithm. The <code>lmdif</code> and <code>lmder</code> routines were written by Burton S. Garbow, Kenneth E. Hillstrom, Jorge J. Moré (Argonne National Laboratory, 1980).</p>
<ul>
<li><strong>Min delta-M/M</strong> - A refinement will stop when the change in the minimization function, <em>M</em>,  <span class="arithmatex">\( (M=\sum[w(I_o-I_c)^2]) \)</span> is less than this value. The allowed range is <span class="arithmatex">\(10^{-9}\)</span> to 1.0, with a default of 0.001. A value of 1.0 stops the refinement after a single cycle. Values less than <span class="arithmatex">\(10^{-4}\)</span> cause refinements to continue even if there is no meaningful improvement.</li>
<li><strong>Max cycles</strong> - This determines the maximum number of refinement cycles that will be performed. This is used only with the "analytical Hessian" and "Hessian SVD" minimizers.</li>
<li><strong>Initial lambda</strong> - Note that here λ is the Marquardt coefficient, where a weight of 1+λ is applied to the diagonal elements of the Hessian. When λ is large, this down-weights the significance of the off-diagonal terms in the Hessian. Thus, when λ is large, the refinement is effectively one of steepest-descents, where correlation between variables is ignored. Note that steepest-descents minimization is typically slow and may not always find the local minimum. This is only used when with the "analytical Hessian" minimizer is selected.</li>
<li><strong>SVD zero tolerance</strong> - This determines the level where SVD considers derivative values to be the same. Default is <span class="arithmatex">\(10^{-6}\)</span>. Make this larger when problems occur due to correlation. This is used only with the "analytical Hessian" and "Hessian SVD" minimizers.</li>
<li><strong>Initial shift factor</strong> - This provides an initial scaling ("damping") for the first cycle of refinement. Only available for "analytic Jacobian" and "numeric" minimizers. </li>
</ul>
</li>
<li>
<p><strong>Single Crystal Refinement Settings</strong>: 
<a name="SingleXtlSettings"></a> 
A set of controls is provided for control of single-crystal refinements. These only appear when single crystal (HKLF) histograms are present in the project.</p>
<ul>
<li><strong>Refine HKLF as F^2?</strong> - When checked, refinements are against <span class="arithmatex">\(F^2\)</span> rather than <span class="arithmatex">\(|F|\)</span>.</li>
<li><strong>Min obs/sig</strong> - Conventional cutoff for single crystal refinements as to what reflections should be considered observed, typical values are 2.0 (<span class="arithmatex">\(2\sigma\)</span>) or 3.0 (<span class="arithmatex">\(3\sigma\)</span>).</li>
<li><strong>Min extinct.</strong> - Reflections with extinction corrections larger than this value are ignored.</li>
<li><strong>Max delt-F/sig</strong> - Removes reflections that are very poorly fit. Should be used only with extreme care, since poorly-fit reflections could be an indication that the structure is wrong.</li>
<li><strong>Max d-spacing</strong> - Reflections with d-space values larger than this value are ignored.</li>
<li><strong>Min d-spacing</strong> - Reflections with d-space values smaller than this value are ignored. </li>
</ul>
</li>
<li>
<p><strong>Sequential Settings</strong>: 
<a name="SequentialSettings"></a> 
 A set of controls is for sequential refinement. Settings here determine if a "normal" or "sequential" refinement is performed. If no datasets are selected here, then all histograms linked to phases in the project and that are flagged as "used" are included in one potentially large (combined) refinement. However, if any number of histograms are selected here as used in a sequential fit, then each of the selected histograms is fitted in turn, which in GSAS-II is called a sequential refinement. Only the first item below is shown in "normal" mode.</p>
<ul>
<li><strong>Select datasets/Reselect Datasets</strong> - This brings up a menu where histograms can be selected, which potentially switches between a normal and a sequential refinement. If one or more histograms are selected, a sequential refinement is used. If none are selected, then the refinement be set as "normal". The button is labeled "Select" when in normal refinement mode and "Reselect" in sequential refinement mode.</li>
<li><strong>Reverse order?</strong> - Normally, in a sequential refinement histograms are fit in the order they are shown in the data tree (which can be reordered by dragging tree items), but when this option is selected, the sequential fit is performed in the opposite order, with the last tree entry first.</li>
<li><strong>Copy results to next histogram?</strong> - When this option is selected, the fitted parameters from each refinement are copied to the next histogram, so that the starting point for each refinement will be the results from fitting the previous. This works well for parametric experiments where parameters such as the lattice parameters change gradually over the course of successive measurements. This option is usually used only for the initial refinement after a sequential fit is started and the setting is cleared once that refinement is completed. For subsequent refinements, it is usually better to start with the results from the previous fit.</li>
<li><strong>Clear previous seq. results</strong> - When this button is pressed, the "Sequential Results" entry with the results from the last sequential fit is deleted from the tree.</li>
</ul>
</li>
<li>
<p><strong>Global Settings</strong>: This is a location for parameters that apply to an entire project. At present there is only one:</p>
<ul>
<li><strong>CIF Author</strong> - The value provided here is used when creating a CIF of an entire project. </li>
</ul>
</li>
</ol>
<H3 style="color:blue;font-size:1.1em">What can I do here?</H3>

<p>This offers a place to change how GSAS-II performs refinements, but has no specific menu commands or graphics.</p>
<p><a name="Covariance"></a></p>
<h2 id="covariance">Covariance</h2>
<p>This window contains residual information after the last refinement. When this tree item is selected, the GSAS-II Plots window shows '<strong>Covariance</strong>', with a graphical representation of the variance-covariance matrix. The text in the data window shows statistical values related to the current fit. At the bottom of this list are buttons that display a horizontal bar chart with the shift to standard uncertainty ratio for each parameter.</p>
<ul>
<li>The button labeled "last refinement" shows these ratios based on the differences between each parameter value from the beginning of the refinement run and the value at the end of the refinement run.</li>
<li>The button labeled "last cycle" shows these ratios based on the differences between each parameter value at the beginning of the last refinement least squares cycle and the value at the end of the cycle (and the refinement run). (These "last cycle" values are only available when the <a href="#Controls">Controls</a> are set for an "analytic Hessian" refinement.) </li>
</ul>
<H3 style="color:blue;font-size:1.1em">What is plotted here?</H3>

<p>The variance-covariance matrix as a color-coded array is shown on this page. The color bar to the right shows the range of covariances (-1 to 1) and corresponding colors. The parameter names are to the right and the parameter numbers are below the plot. </p>
<H3 style="color:blue;font-size:1.1em">What can I do with the plot?</H3>

<ul>
<li>Move the mouse cursor across the plot. If on a diagonal cell, the parameter name, value and standard uncertainty (esd) is shown both as a tooltip and in the right-hand portion of the status bar. If the cursor is off the diagonal, the two parameter names and their covariance are shown in the tool tip and the status bar.</li>
<li>Use the Zoom and Pan buttons to focus on some section of the variance-covariance matrix.</li>
<li>Press 's' – A color scheme selection dialog is shown. Select a color scheme and press OK, the new color scheme will be plotted. The default is 'RdYlGn'.</li>
<li>Press 'p' – Saves the covariance values in a text file.</li>
</ul>
<p><a name="Constraints"></a></p>
<h2 id="constraints">Constraints</h2>
<p>This window shows the constraints to be used in a refinement. Constraints are divided with a tab for each type: Phase, Histogram/Phase, Histogram, Global and Sym-Generated. Note that the standard parameters in GSAS-II are divided into three classes and appear respectively on the Phase, Histogram and Histogram/Phase tabs:</p>
<ul>
<li>those pertaining to quantities in each phase (naming pattern <code>p::name</code>); examples include atom coordinates, thermal motion and site fraction parameters;</li>
<li>those pertaining to quantities in each histogram (naming pattern <code>:h:name</code>); such parameters are those that depend only on the data set: the scale factor and profile coefficients (e.g. U, V, W, X and Y);</li>
<li>and those pertaining to quantities defined for each histogram in each phase (naming pattern <code>p:h:name</code>); these parameters are quantities that can be dependent on both the phase properties and the sample or dataset used for the measurement. Examples include phase fractions and sample-broadening coefficients such as microstrain and crystallite size; they are found in the Data tab for each phase.</li>
</ul>
<p>The following types of constraints may be specified by users:</p>
<ul>
<li>
<p><strong>Holds</strong> - Use this to prevent a parameter from being refined. Most valuable when refinement of a parameter is selected in a group for refinement (such as x, y &amp; z for an atom or unit cell parameters) and one must be fixed. For example, if the space group for a phase has a polar axis (e.g., the y-axis in monoclinic space group <span class="arithmatex">\(P \ \rm 2_1\)</span>, the origin with respect to <em>b</em> is arbitrary). With a polar axis present, if an arbitrary constant is added to all atom coordinates along that axis, a completely equivalent structure is obtained. Thus in <span class="arithmatex">\(P \ \rm 2_1\)</span> it is not possible to refine the y coordinates for all atoms. Place a Hold on any one atom y coordinate to keep the structure from drifting up or down the y-axis during refinement.</p>
</li>
<li>
<p><strong>Equivalence assignments</strong> - Determines a set of parameters that should have values with a specified ratio. Atom coordinates are handled slightly different, where the ratios are specified for the applied shifts, not the actual coordinate values. Examples for typical use are sets of atoms that should be constrained to have the same displacement parameters (aka thermal motion, Uiso, etc.) or sets of profile coefficients U, V &amp; W across multiple data sets. Note that the first selected parameter is treated as independent, and the remainder are "slaved" to that parameter as "dependent parameters." All parameters in an equivalence must be varied. If any parameter is not varied or is given a "hold," a warning is displayed and none of the parameters in the equivalence are refined.</p>
</li>
<li>
<p><strong>Constraint Equations</strong> - Defines a set of parameters whose sum (with possible non-unitary multipliers) will be equal to a constant. For example, a common use for this is to specify the sum of occupancies for atoms sharing a site have a sum fixed to unity or so that the sum of occupancies for an atom type that is occurs on several sites is fixed to match a composition-determined value. Note that all parameters in the equation are considered as "dependent parameters." If a parameter in a constraint equation is held or is not varied, that parameter is removed from the equation (the sum value is modified accordingly). If no parameters remain the equation is ignored.</p>
</li>
<li>
<p><strong>New Var assignment</strong> - These are similar to constraint equations in that they define a set of parameters and multipliers, but rather than specifying a value for the expression, a new parameter is assigned to that sum. Thus these constraints create new variable that effectively replace the original GSAS-II parameters with ones supplied by the user. This is done by replacing a degree of freedom from the original variables with a new one that modifies the parameters, where the shift is applied according to the ratio specified in the expression. This can be used to create new parameters that redefine the relationships between items such as coordinates or magnetic moments. The new parameter may optionally be named by the user. The new var expression creates a new global parameter, where that new parameter is independent, while all the parameters in the expression are considered as dependent. The setting of the refine flags for the dependent parameters is not used. Only if the new var parameter is marked as <code>refine</code> then it will be refined. However, if any dependent variable is set as "hold," the <strong>new var</strong> parameter will not be refined.</p>
<p><strong>New Var</strong> constraints are generated when <a href="https://iso.byu.edu/isodistort.php">ISODISTORT</a> is used to develop mode distortions from a comparison of a high symmetry parent structure (e.g. cubic perovskite) with a distorted child substructure. They are implemented for the phase when a special CIF file produced by ISODISTORT from a mode distortion analysis is imported. You can select which distortion modes to refine. Refining all modes is equivalent to refining the atoms freely and then importing the resulting structure to ISODISTORT to determine the mode amplitudes. </p>
</li>
</ul>
<p>Note that when new var and constraint equation constraints are defined, they create new global parameters. Constraints on these will be rare, but can be managed on the Globals tab. Finally, some constraints are defined automatically based on restrictions determined by space group symmetry. These constraints can be seen, but not changed, using the Sym-Generated tab. Other constraints (holds) will be created when rigid bodies are specified.</p>
<H3 style="color:blue;font-size:1.1em">What can I do here?</H3>

<p>Select the tab for the parameter type(s) you wish to constrain then create new parameters using the "Edit Constr." menu commands:</p>
<ul>
<li><strong>Add Hold</strong> - Select a parameter that you wish to remain fixed. If selected, a dialog box will appear showing the list of available parameters; select one and then OK to implement it, Cancel will cancel the operation. The held parameter will be shown in the constraint window with the keyword 'FIXED'.</li>
<li><strong>Add equivalence</strong> - Select the independent parameters and press OK; a second dialog box will appear with only those parameters that can be made equivalent to the first one. Choose those and press OK. Cancel in either dialog will cancel the operation. The equivalenced parameters will show as an equation of the form <span class="arithmatex">\(M1*P1+M2*P2=0\)</span>; usually M1=1.0 and M2=-1.0 but can be changed via the 'Edit' button. The equation(s) are shown in the window tagged by 'EQUIV' to mark it as an equivalence assignment.</li>
<li><strong>Add constraint</strong> - If selected, a dialog box will appear with a list of the available parameters. Select one and press OK; a second dialog box will appear with only those parameters that can be used in a constraint with the first one. Choose those and press OK. Cancel in either dialog will cancel the operation. The equivalenced parameters will show as an equation of the form <span class="arithmatex">\(M1*P1+M2*P2+…=C\)</span>; the multipliers M1, M2, … and C can be changed via the 'Edit' button. The equation is shown in the window tagged by 'CONSTR' to mark it as a constraint equation assignment.</li>
<li><strong>Add New Var</strong> - This behaves very much like the "Add constraint" menu command except that it defines a new parameter rather than define a value for the expression. That new var parameter can optionally have a named assigned. The expression is displayed with the keyword 'New Var' to mark its type. Note that a 'Refine?' box is included for this type of constraint.</li>
<li><strong>Make atoms equivalent</strong> - This provides a shortcut for establishing constraints when two share a single site. Coordinates and Uiso values are constrained to be the same and site fractions are constrained to add to 1.</li>
<li><strong>Show ISODISTORT modes</strong> - Used after a CIF from the ISODISTORT web site is read, which will display the values for the normal modes from representational analysis from the coordinates.</li>
</ul>
<p>In addition to menu commands, this window also offer the following actions by pressing buttons:</p>
<ul>
<li><strong>Show Errors</strong> - this button will be active if serious errors -- that would prevent a refinement from being performed -- are encountered processing the constraints.</li>
<li><strong>Show Warnings</strong> - this button will be active if correctable problems are encountered in processing the constraints, such as a constraint being rejected because a parameter is not varied. These warning may indicate that the choice of which parameters will be refined is not what was planned.</li>
<li><strong>Show Generated Constraints</strong> - After constraints have been processed, a series of relationships are developed to determine new variables from the current parameters and "inverse" equations that determine dependent parameters from the new variables and independent parameters. This shows the resulting relationships, as well as any "Hold" variables.</li>
<li><strong>Delete Selected</strong> - This button will cause all the selected constraints on the current tab to be deleted.</li>
</ul>
<p><a name=Constraints-SeqRef></a></p>
<h3 id="sequential-refinement-constraints">Sequential Refinement Constraints</h3>
<p>While all the general information on constraints (above) applies to sequential refinements, the sequential refinement is performed by fitting each histogram individually and this affects how constraints are defined and processed for parameters keyed to a particular histogram number. When sequential refinement is selected (via the <a href="#Controls">Controls</a> tree item), it becomes possible to define constraints of form <code>p:*:name</code> and <code>:*:name</code> (where "p" is a phase number and name is a parameter name). The "*" here is called a wildcard, and in a constraint or equivalence will cause that to be used for every histogram in turn.</p>
<p>In sequential refinement mode, two additional controls are shown in the Constraints window. The first, which is labeled wildcard use, specifies what is done when a specific histogram is referenced by number in a constraint or equivalence by number rather than by wildcard. This offers three modes:</p>
<ul>
<li><strong>Set hist # to *</strong> - Any constraints previously specified with a specific histogram number will be changed to apply to all histograms. This is the default for new projects.</li>
<li><strong>Ignore unless hist=*</strong> - Any constraints previously specified with a specific histogram number will be ignored and constraints with a "*" for a histogram number will be used. Note that constraints on phase parameters (of form 'p::name' -- without a histogram number specified) will be used normally. Note that this was the operating mode for GSAS-II in earlier versions before these buttons were introduced.</li>
<li><strong>Use as supplied</strong> - If different constraints are to be applied to different histograms, it becomes necessary to create constraints with specific histogram numbers. In this mode, constraints specified with a specific histogram are applied only to that histogram, while wildcarded ones are applied to all histograms. Note that one should not specify two constraints on a single parameter, one with a wildcard and one with a specific histogram number, as both will be applied to the specified histogram which results in an unsatisfiable conflict.</li>
</ul>
<p>Also included when sequential refinement is selected is a menu button labeled "Selected histogram." With this it is possible to look at constraint problems when processing a specific histogram.</p>
<p><a name="Restraints"></a></p>
<h2 id="restraints">Restraints</h2>
<p>This window shows the restraints to be used in a refinement for each phase (if more than one). It is organized into several tabbed pages, one page for each type of restraint. Restraints are developed for an individual phase and act as additional observations to be "fitted" during the refinement. </p>
<p>Note that a restraint "pushes" a refinement towards a target value, but does not require that to happen (unlike a constraint). The strength of the "push" is dependent on the weighting factor and the "esd" used on the target value. The ideal use of a restraint is for an aspect of a structure that is not well-determined by the data. In this case, when the weighting of a restraint is dropped, the value of the restrained parameter may move considerably from the target value, but the quality of the fit (noted by the Rietveld plot, R-factor, reduced <span class="arithmatex">\(\chi^2\)</span>, etc.) is not expected to change very much. If one places a restraint that is not consistent with the model (<em>e.g.</em> is wrong), then the fit will improve significantly as the weight is decreased. Such feedback is not possible with a constraint.  <a href="#RestraintWeights">See below</a> for more discussion on restraint weighting. </p>
<H3 style="color:blue;font-size:1.1em">What can I do here?</H3>

<ul>
<li>Select the tab for the restraint type you wish to use. Each will have the same possibilities in the 'Edit' menu.</li>
<li>You can change the Restraint weight factor – this is used to scale the weights for the entire set of restraints of this type. Default value for the weight factor is 1.0.</li>
<li>You can choose to use or not use the restraints in subsequent refinements. Default is to use the restraints.</li>
<li>You can change the search range used to find the bonds/angles that meet your criteria for restraint.</li>
<li>You can examine the table of restraints and change individual values; grayed out regions cannot be changed. The 'calc' values are determined from the atom positions in your structure, 'obs' values are the target values for the restraint and 'esd' is the uncertainty used to weight the restraint in the refinement (multiplied by the weight factor).</li>
<li>
<p>Menu <strong>Edit</strong> – has the menu commands listed below. Note that some entries may be grayed out, if not appropriate for your phase or for the selected restraint.</p>
<ul>
<li><strong>Add restraints</strong> - this takes you through a sequence of dialog boxes which ask for the identities of the atoms involved in the restraint and the value to be assigned to the restraint. The esd is given a default value which can be changed after the restraints are created.</li>
<li>
<p><strong>Add residue restraints</strong> - if the phase is a 'macromolecule' then develop the restraints from a selected 'macro' file based on those used in GSAS for this purpose. A file dialog box is shown directed to /GSASIImacros; be sure to select the correct file.</p>
</li>
<li>
<p><strong>Add MOGUL restraints</strong> - add restraints from a file produced by the CCDC's MOGUL program or a file prepared 
in that style. <a href="#RestraintImports">See below</a> for more information on importing restraints. 
Note that when used from the Bond tab, only "bond" entries are read and when used from the 
Angle tab, only "angle" entries are read.</p>
</li>
<li>
<p><strong>Plot residue restraints</strong> - if the phase is a 'macromolecule' and the restraint type is either 'Torsion restraints' or 'Ramachandran restraints', then a plot will be made of the restraint distribution; torsions as 1-D plots of angle vs. pseudopotential energy and Ramachandran ones as 2-D plot of psi vs phi. In each case a dialog box will appear asking for the residue types or specific torsion angles to plot. Each plot will show the observed distribution (blue) obtained from a wide variety of high-resolution protein structures and those found (red dots) for your structure. The restraints are based on a pseudopotential (red curve or contours – favorable values at the peaks) which has been developed from the observed distributions for each residue type.</p>
</li>
<li><strong>Change value</strong> - this changes the 'obsd' value for selected restraints; a dialog box will appear asking for the new value.</li>
<li><strong>Change esd</strong> - this changes the 'esd' value for selected restraints; a dialog box will appear asking for the new value.</li>
<li><strong>Delete restraints</strong> - this deletes selected restraints from the list. A single click in the blank box in the upper left corner of the table will select/deselect all restraints.</li>
</ul>
</li>
</ul>
<h3 id="restraint-types">Restraint types</h3>
<ul>
<li><strong>Bond</strong>: Defines restraints with a target distance between pairs of atoms (which are typically bonded, but this is not required). Atom pairs to be restrained are set by performing a search and then selecting the pairs to be used. </li>
<li><strong>Angle</strong>:  Defines restraints with a target angle for a set of three atoms. Atoms to be used are located via a search and selection. </li>
<li><strong>Plane</strong>: Used to restrain selected atoms to be approximately co-planar. This is most commonly used for macromolecular crystallography. </li>
<li><strong>Chiral</strong>: Defines a chiral volume restraint. This is unlikely to be used other than for macromolecular crystallography. </li>
<li><strong>Chem. Comp.</strong>: This is of use when "Frac" parameters (atom occupancies) are refined. This option allows a variety of different types of restraints to be created. More than one chemical composition restraint can be defined for a phase. Examples of how this might be used would be to "push" refinement towards an expected composition, to encourage charge balance or to conserve valences. </li>
<li><strong>General</strong>: This allows a quantity to be restrained that is computed from a user-supplied equation, based on GSAS-II parameters. Thus, one can create any type of restraint that is desired. This restraint is used by supplying a Python equation and then to define which GSAS-II parameter is associated with each variable in the Python equation. One also supplies the target value for the restraint. Note that it is possible to use externally defined functions that contain "if" statements, which allows restraints that enforce minimum or maximum quantities rather than target values to be defined. </li>
</ul>
<p><a name="RestraintImports"></a></p>
<h3 id="importing-restraints">Importing Restraints</h3>
<p>For the  <strong>Add MOGUL restraints</strong> menu command, restraints are read from a supplied file. 
The file produced by the MOGUL program has many columns, but only the first 8 are used here and a few of those do not matter. 
If producing the file manually or with your own software, follow the following instructions. 
The first line in the file must be a header, which must start with "Type" (note capitalization). 
The rest of the line is ignored, but it is convenient to follow the example below. </p>
<p>Subsequent lines should contain:</p>
<ol>
<li>The first column (Type) should be "bond" or "angle"</li>
<li>The second column (Molecule)  is ignored. It can be blank, but a comma to end the field is required</li>
<li>The third column (Fragment) specifies the atoms in the distance or angle. 
   The atoms should be separated by at least one space. For an angle the second listed atom is the one 
   at the apex of the angle. Capitalization of the atom name must match the name used in the atoms table. </li>
<li>The fourth column (Classification) can be anything other than "No hits" </li>
<li>The fifth column (No. of hits) is ignored. It can be blank, but a comma to end the field is required</li>
<li>The sixth column (Query value) is ignored. It can be blank, but a comma to end the field is required</li>
<li>The seventh column (Mean) is the bond distance or bond angle used in the restraint, labeled "target" in the table.</li>
<li>The eighth column (Std. dev.) is the uncertainty on the bond distance or bond angle labeled "esd" in the table. 
   Note that quantity <code>"Weight factor" * (obs-calc) / esd</code> determines the impact of the restraint, 
   so the esd values are on an arbitrary scale. </li>
</ol>
<p>If the columns are reorganized and the header changed to match, the code should be able to change to match, but this has not been tested. 
Spacing between columns does not matter and columns do not need to be aligned. 
With a file such as the one below, that contains both distances and angles, the file must be 
read in twice, using the <strong>Add MOGUL restraints</strong> menu command from both the Bond tab and 
the Angle tab to read in both sets of restraints. Blank lines are ignored. Note that only restraints
between atoms within the asymmetric unit can be generated with this type of file. If symmetry or translations are 
needed to generate any of the atoms in the distance or angle, the <strong>Add restraints</strong> menu command must
be used to search.</p>
<p>This is an example file to be used to read in restraints:</p>
<pre><code>Type, Molecule,    Fragment,  Classification, No. of hits, Query value,  Mean,   Std. dev.
bond,   ignore,    Co1 O16,     OK,           ignore,      ignore,      2.060,     0.001
bond,   ignore,    Co1 O20,     OK,           ignore,      ignore,      2.069,     0.001

angle,  ignore,    N15 C8 C9,   OK,           ignore,      ignore,      123.38,    0.01
angle,  ignore,    C9 C10 C11,  OK,           ignore,      ignore,      116.18,    0.01
angle,  ignore,    C11 C12 N15, OK,           ignore,      ignore,      123.91,    0.02
</code></pre>
<p><a name="RestraintWeights"></a></p>
<h3 id="restraint-weights">Restraint Weights</h3>
<p>There is no simple recipe for how to weight restraints, in that it varies with the problem, but one should understand is that restraints are used computationally just like any other data point, where the refinement works towards minimizing the sum of 
<span class="arithmatex">\([ w * (obs_i - calc_i ) / \sigma_i  ]^2\)</span> where w is usually 1 and <span class="arithmatex">\(obs_i\)</span> might be a powder diffraction intensity but could also be a restraint bond distance and where <span class="arithmatex">\(\sigma_i\)</span> is the uncertainty on <span class="arithmatex">\(obs_i\)</span>, which for restraints is the "esd" value. Since there are usually <span class="arithmatex">\(10^3\)</span> to <span class="arithmatex">\(10^4\)</span> data points and only circa <span class="arithmatex">\(10^1\)</span> restraints, I will usually initially weight the restraints quite highly (100-10,000) since I want to start the fit with a model that closely matches the expected distances and angles. </p>
<p>As the refinement progresses and I have a good fit, I will then try lowering the restraint weighting. Ideally, I can set the weight on the distances &amp; angles to zero, in which case the restraints will show me a diagnostic of how far the as-fit distances, etc. deviate from my presupposed values, but do not actually change the fit. </p>
<p>A weight factor of 0 may not produce a good result with complex structures and not very complex diffraction patterns. In those cases, there just is not enough information to define all the structural degrees of freedom uniquely. In that case, lowering the weighting will increase the deviations seen in the restraints with only very trivial improvements in the fit to the powder pattern. One needs to reduce the restraint weight, but only to the point where a chemically reasonable structure is still obtained. Usually, a small value for the weight is needed (around 0.1 to 10) for the final refinements, as very little "force" is needed to keep the structure in agreement with what is expected for the class of materials. </p>
<p>There is one important case to be aware of, and that is where the restraints are incorrect for the material being studied. If one tries to restrain, for example, a bond to a value that is wrong for the material (for example a carbon-carbon single bond where in fact a double-bond is present) and the data have enough sensitivity to determine this, then the restraints and the data will be in conflict. Lowering the restraint weighting will produce a significant improvement in the fit to the powder data as the bond distance refines away from the target value. If this occurs the restraints need to be corrected to the valid structure. This is different from the case where there is not enough information to define all the structural degrees of freedom, where the fit to the powder data improves only slightly as the restraint is down-weighted. In that latter case, there are very many structures that produce about the same quality of fit, but restraints will prejudice the fit towards models that are more expected based on what is known on the class of materials. </p>
<p><a name="Rigidbodies"></a>
<a name="Rigid_bodies"></a></p>
<h2 id="rigid-bodies">Rigid bodies</h2>
<p>There are three different types of rigid bodies that can be used in GSAS-II, as selected by the tabs at the top of this window. </p>
<ul>
<li><strong>vector rigid bodies</strong>, where atoms are defined in terms of multiple displacements from an origin. </li>
<li><strong>residue rigid bodies</strong>, where atoms are defined according to Cartesian coordinates and torsion angles. These are much more commonly used than vector rigid bodies. </li>
<li><strong>spinning rigid bodies</strong>, where the dynamics or disorder causes the atoms to not have specific locations in the unit cell. </li>
</ul>
<p>Note that there are two steps in defining a rigid body. In this data item the rigid body is defined. The rigid body is then later inserted into one or more phases using the <a href="phaseRB.html">"RB Models"</a> tab on a Phase data item. A rigid body can be inserted into more than one phase. </p>
<H3 style="color:blue;font-size:1.1em">What can I do here?</H3>

<ul>
<li>Select the tab for the rigid body type you wish to use. </li>
<li>Once the rigid body type has been selected, use the menu commands listed below. The menu contents will depend on which type of rigid body is selected.</li>
<li>
<p>Editing of some of much of the rigid body information can be done in the data window once a body is created. </p>
<ul>
<li>
<p><strong>Add rigid body</strong> - (Vector &amp; Spinning rigid bodies). For vector bodies this creates a vector description of a rigid body. A dialog box asks the number of atoms (&gt;2) and the number of vectors required to create the rigid body. An entry will be created showing a magnitude with the vector set to be applied for each vector needed to develop the rigid body.</p>
</li>
<li>
<p><strong>Extract from file</strong> - (Vector &amp; residue rigid bodies). This prepares a rigid body by reading in coordinates (typically as fractional coordinates) and then atoms are selected from the file that is read and finally the origin and axes of the Cartesian system used by the rigid body are defined.</p>
</li>
<li>
<p><strong>Save rigid body</strong> - (Vector &amp; Spinning rigid bodies). Writes the contents of a defined rigid body into a file where it can be reused in another GSAS-II project. A different format is used for vector vs. residue bodies.</p>
</li>
<li>
<p><strong>Read rigid body</strong> - (Vector &amp; Spinning rigid bodies). Creates a rigid body from a file written perviously by a "Save rigid body" command.</p>
</li>
<li>
<p><strong>Add translation</strong> - (Vector rigid bodies). The Cartesian coordinates for the body are created from the sum of coordinate matrices multiplied by a value that converts the arbitrary axis lengths to Angstroms. One can sum 9 multiplier/coordinate matrix sets. This command will add an additional "translation" consisting of a multiplier and coordinates for every atom in the rigid body. </p>
</li>
<li>
<p><strong>Import XYZ</strong> - (Residue rigid bodies) this reads a text file containing a set of Cartesian coordinates describing a rigid body model. Each line has atom type (e.g. C, Na, etc.) and Cartesian X, Y and Z.</p>
</li>
<li>
<p><strong>Save as PDB</strong> - (Vector rigid bodies). Writes the contents of a defined rigid body as Cartesian coordinates in a PDB file. </p>
</li>
<li>
<p><strong>Define Torsion</strong> - (Residue rigid bodies) this adds a variable torsion angle in an existing rigid body via a sequence of dialog boxes. The first one asks for the origin and the second asks for the pivot atom for the torsion from the nearest neighbors to the origin atom; the atoms that ride on the selected torsion are automatically found from their bond lengths.</p>
</li>
<li><strong>Import residues</strong> - (Residue rigid bodies) this reads a predetermined macro file that contains standard (Engh &amp; Huber) coordinates for the amino acids found in natural proteins along with predetermined variable torsion angle definitions.</li>
</ul>
</li>
<li>
<p>Once a rigid body is defined you can plot it, change its name or manipulate any torsion angle to see the effect on the plot.</p>
</li>
<li>The translation magnitudes in a vector rigid body can be refined allowing some lengths within a vector rigid body to be varied.</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="javascripts/mathjax.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>